File: ChessEngine/chessEngineCPP/knight.cpp
#include "knight.h"
#include <map>
#include <vector>

// Knight move offsets
const std::pair<int, int> knight_offsets[8] = {
    {1, 2}, {1, -2}, {-1, 2}, {-1, -2},
    {2, 1}, {2, -1}, {-2, 1}, {-2, -1}
};

void check_knight(const std::map<std::pair<int, int>, char>& current_board, 
                  std::map<char, std::vector<std::pair<std::pair<int, int>, std::pair<int, int>>>>& moves_list) 
{
    for (const auto& piece : current_board) {
        std::pair<int, int> position = piece.first;
        char piece_type = piece.second;
        
        if (piece_type == 'N') {
            // White knight moves
            for (const auto& offset : knight_offsets) {
                std::pair<int, int> new_pos = {position.first + offset.first, position.second + offset.second};
                if (current_board.find(new_pos) != current_board.end() && 
                    (islower(current_board.at(new_pos)) || current_board.at(new_pos) == ' ')) 
                {
                    moves_list['b'].emplace_back(position, new_pos);
                }
            }
        } 
        else if (piece_type == 'n') {
            // Black knight moves
            for (const auto& offset : knight_offsets) {
                std::pair<int, int> new_pos = {position.first + offset.first, position.second + offset.second};
                if (current_board.find(new_pos) != current_board.end() && 
                    (isupper(current_board.at(new_pos)) || current_board.at(new_pos) == ' ')) 
                {
                    moves_list['w'].emplace_back(position, new_pos);
                }
            }
        }
    }
}

File: ChessEngine/knight.py
knight_offets = [(1, 2), (1, -2), (-1, 2), (-1, -2), (2, 1), (2, -1), (-2, 1), (-2, -1)]

def check_knight(current_board: dict, moves_list: list):
    """
    The function `check_knight` iterates through a chess board to find valid knight moves for white and
    black knights.
    
    :param current_board: The function `check_knight` takes a current chess board represented as a
    dictionary where keys are positions on the board and values are the pieces at those positions, and a
    list of moves. It checks for valid knight moves on the board and appends them to the moves list
    :type current_board: dict
    :param moves_list: It seems like the definition of the `moves_list` parameter is missing in your
    code snippet. Could you please provide more information or context about the `moves_list` parameter
    so that I can assist you further with the `check_knight` function?
    :type moves_list: list
    """
    for k,v in current_board.items():
        if v == "N":
            for offset in knight_offets:
                if (k[0] + offset[0], k[1] + offset[1]) in current_board and (current_board[(k[0] + offset[0], k[1] + offset[1])].islower() or current_board[(k[0] + offset[0], k[1] + offset[1])] == ""):
                    moves_list['b'].append((k, (k[0] + offset[0], k[1] + offset[1])))
        if v == "n":
            for offset in knight_offets:
                if (k[0] + offset[0], k[1] + offset[1]) in current_board and (current_board[(k[0] + offset[0], k[1] + offset[1])].isupper() or current_board[(k[0] + offset[0], k[1] + offset[1])] == ""):
                    moves_list['w'].append((k, (k[0] + offset[0], k[1] + offset[1])))